### DISCUSSION TITLE
Generate a ton of themes based on editor themes

### DISCUSSION THREAD

**Discussion by corysimmons** [jln13x/ui.jln.dev#9] [Ideas]
This is just a fun thing, but it'd be nice to generate a bunch of classic themes (from terminals and editors) as a themes on the app.

Like I'm curious what Solarized or Material Theme would look like as shadcn-ui themes.

Obviously, we wouldn't have to match up all the colors completely, but just finding some pattern like `primary`, `accent`, `background`, etc. and filling those in to work would be interesting.

Again, just a silly feature. :)

**Comment by jln13x** [comment]
I was thinking about that aswell. Do you have any idea where we could find all the themes? Instead of having to search for them one by one...

But ye, we would just have to figure out their data structure (e.g. vscode theme) and map them to the shadcn structure

**Reply by corysimmons** [reply]
- https://github.com/dennis84/codemirror-themes/tree/main/theme 
- https://github.com/brijeshb42/monaco-themes/tree/master/themes
- https://github.com/FarhadG/code-mirror-themes/tree/master/themes

This vscodethemes.com [repo](https://github.com/vscodethemes/web) has a ton of nice code for fetching popular themes directly from the VSCode Extensions Marketplace, but I don't want to set up Cloudflare workers and I'm not comfy with Remix, so it seems a bit complicated (wish he just had some `$ node fetch-themes-and-put-them-in-a-file` util 😢).

I'm trying to coerce ChatGPT into making such a util right now, but it seems like you need to:
1. Fetch the popular themes from VSCode Extension Marketplace with Puppeteer.
2. Nav to their GitHub repo.
3. Search for something like `editor.background`, then find all the .json files that are returned from that (usually named something like `Solarized.json`, `Solarized-Dark.json`, etc.)
4. Then convert all the `editor.background` and such to `background`, `foreground`, etc.

I'll tinker with it for a bit and let you know if I figure it out.

**Reply by corysimmons** [reply]
I got a Puppeteer script working, but almost every single one of the repos on https://marketplace.visualstudio.com/search?target=VSCode&category=Themes&sortBy=Installs generates their theme in a really complicated way.

Like some of them have all their colors in a YAML file. Some of them are in JSON5 files. Some of them generate VSCode theme property names from their own property names (e.g. https://github.com/dracula/visual-studio-code/blob/e475d548db27773fa0004b252c0a4701f187fb7e/src/dracula.yml `&BG` is converted into `editor.background` on build, or https://github.com/material-theme/vsc-material-theme/blob/9c1313579cecdd377fc475ccf108da6402216745/scripts/generator/color-set.ts#L300), etc. etc. etc.

Yet _another_ example of devs not publishing their `dist` to GitHub (even when the dist is tiny), and it causing problems. 😑

Another approach might be:
1. Make a Docker image with VSCode.
2. Have it [install](https://github.com/jln13x/ui.jln.dev/assets/224636/ab0829d6-ce9d-4028-8529-59a2b7f68593) all the top themes somehow.
3. Then I guess [loop](https://github.com/jln13x/ui.jln.dev/assets/224636/1f3cbc75-b169-492a-8542-2b5177494a7b) over dirs in `.vscode/extensions` for anything with `theme` in the filename.
4. Look at the `package.json` to get `contributes.themes` [JSON filepaths](https://github.com/jln13x/ui.jln.dev/assets/224636/6470bcac-0ae9-44df-851a-24953ffae9ce) which actually have predictable color properties.
5. Look at those files to generate a `themes.json` we could work with to match up styles.

Another, another, approach might be to just install the themes locally, copy/paste the JSON to a bigger JSON, uninstall the themes. If we're just talking about 10 or so, it wouldn't take long too long.

But honestly, I don't think it's worth the effort. It was just a idea after all--who knows if those color schemes would even translate well to shadcn-ui.

I'll open a ticket on vscodethemes.com in hopes they'll expose a function to let us use their work. 😄 https://github.com/vscodethemes/web/issues/269

Lemme know if you're super interested in this feature and I'll do the Docker thing.

**Reply by jln13x** [reply]
The data export would be awesome 🔥

**Reply by corysimmons** [reply]
The VSCodeThemes.com maintainer emailed me these CSVs! <3

`Theme.csv` seems to have all the colors in it. `Extension.csv` has data like the amount of downloads and such.

[Extension.csv](https://github.com/jln13x/ui.jln.dev/files/14212957/Extension.csv)
[Theme.csv](https://github.com/jln13x/ui.jln.dev/files/14212958/Theme.csv)

**Reply by jln13x** [reply]
Working on smth 🚀

**Reply by jln13x** [reply]
Feels like the `Theme.csv` is incomplete.

Those are all the colors I get for `Catppuccin Latte`:
![CleanShot 2024-02-08 at 21 43 09@2x](https://github.com/jln13x/ui.jln.dev/assets/85513960/4101cb96-bc5b-4d98-ad08-b26547f31ad4)

Cant seem to find this color
![CleanShot 2024-02-08 at 21 45 20@2x](https://github.com/jln13x/ui.jln.dev/assets/85513960/63507ca7-f69b-45f1-b5a3-1186f65f7022)

Any idea? @corysimmons

**Reply by corysimmons** [reply]
Let me ping @jschr here. 👋 

Did something get broken during the export?

**Reply by corysimmons** [reply]
@jln13x Are you sure this is the CSV and not your code (push a branch?)? I see a lot of colors in Catppuccin Latte:

![image](https://github.com/jln13x/ui.jln.dev/assets/224636/0a560241-8ede-4778-87f4-6330743aba31)

**Reply by jln13x** [reply]
> @jln13x Are you sure this is the CSV and not your code (push a branch?)? I see a lot of colors in Catppuccin Latte:
> 
> ![image](https://github.com/jln13x/ui.jln.dev/assets/224636/0a560241-8ede-4778-87f4-6330743aba31)

There are only 4 unique colors (the one from my screenshot) and the accent color (this pinkish color) is nowhere to be found

**Reply by jschr** [reply]
hey, sorry for the less then great data. I've had this thing running for years and I've had a rewrite on my todo list for a long time now. 

It looks like the data I exported from the main db is not in sync with what I'm storing in Cloudflare KV. I'm not sure why yet. 

I'm going to see if I can export everything from Cloudflare KV, which should give you better data. 

Here's an example of what the export looks like: https://gist.github.com/jschr/db361395272718a0b28ad46508b28c02 

Would this work?

**Reply by jschr** [reply]
I just pushed a really quick feature that might help. 

You can get the JSON of any theme or extension by adding `.json` to the URL. 

For example: 
https://vscodethemes.com/e/catppuccin.catppuccin-vsc/catppuccin-latte.json (for a specific theme in an extension) 
https://vscodethemes.com/e/catppuccin.catppuccin-vsc.json (for all themes in an extension)

I'm currently only syncing the theme values that I use to render the previews on the site so there's a number of missing ones (like the notification badge colors).

Hopefully this helps but let me know if there's anything I can do.

**Reply by corysimmons** [reply]
Here's an array of all those URLs: https://gist.github.com/corysimmons/634b36a67d7af2b7a103840058519171

Currently working on converting it to something a bit cleaner. Will update...

**Reply by corysimmons** [reply]
@jln13x Here's all the theme data from vscodethemes.com https://gist.github.com/corysimmons/4c704eb0ddf1c71e4ef2e73e1cd9a26d

There's a ton of them. I'm not sure if this should be some special "official" collection on the site or what, but I'd love to be able to flip through these once they're live.

Let me know if that works for you! Thanks again to both of you. Really excited to see how these translate to web.

**Reply by jln13x** [reply]
If we are going to add it, it should be a new tab here imo

![CleanShot 2024-02-09 at 10 23 48@2x](https://github.com/jln13x/ui.jln.dev/assets/85513960/0743be39-9a5a-4ac7-955c-b9ea5d5f901f)

**Reply by jln13x** [reply]
The new export seems to be better, will have a look into it on the weekend. Thanks alot you guys, will keep you updated 🔥

**Reply by corysimmons** [reply]
Awesome! I am now also wondering if it might be worth just making a route to allow users with Auth.js tokens to create their own themes.

It might be easier than you manually having to do this, and it opens up the door for devs to generate themes from all kinds of stuff (e.g. browser extension to import a theme from those sites like [Coolors](https://coolors.co/ffc6d9-ffe1c6-fff7ae-48284a-916c80) might be trivial, etc.).

```
POST https://ui.jln.dev/api/v1/themes
{
  headers: { Authorization: Bearer <TOKEN_I_CAN_GET_FROM_/account> }
  body: {
    name: "My Cool Theme",
    light: {
      background: '#000',
      ...
    },
    dark: {
      ...
    }
  }
}
```

I don't know tRPC and haven't written an external API route like this, but I suspect it won't be terribly difficult and I'd like to know how to do it so I'd be willing to work on this if you like the idea.

**Reply by jln13x** [reply]
Would have a lot of implications to consider tho. Such as spam prevention. 
I think I will import by myself for now.

**Reply by jln13x** [reply]
Got a first version working btw, will release it next week.
Feel free to send me your Twitter / X Handle so I can give you guys credits 🙌
@corysimmons  @jschr 

![CleanShot 2024-02-10 at 12 44 04@2x](https://github.com/jln13x/ui.jln.dev/assets/85513960/100f6b19-bc46-4ad8-9bc7-7319d407c703)

![CleanShot 2024-02-10 at 12 45 09@2x](https://github.com/jln13x/ui.jln.dev/assets/85513960/3082d998-814d-4974-b4cf-5c3df83c11a9)

**Reply by jschr** [reply]
This looks great! Looking forward to trying it out. My handle is https://twitter.com/_jschr. Happy I could help!

**Reply by corysimmons** [reply]
Oh wow! Those look better than I thought they would. I haven't done X since I was a teen. 🛼

**Reply by jln13x** [reply]
@corysimmons its live btw 🎉

**Reply by corysimmons** [reply]
Danggg they look good! Will definitely be using these! So happy they turned out well!

**Comment by corysimmons** [comment]
@jln13x I know you're probably burnt out on this feature, but what are your thoughts on just inverting the colors to provide an opposite theme?

That is all the shadcn-ui themes have a dark + light theme, but all these VSCode themes only provide one.

Do they look good if you just invert them? (you'd have to set which one is the dark theme based on something like the background color)

**Reply by jln13x** [reply]
Should be possible, also played around with a bit already.  Just wanted to get this feature out quickly.

Gotta be lots of trial n error tho, maybe next weekend 👀
